Enable multi circuit submission in PennyLane IonQ, second attempt.

radumarg opened the pull request:
Enable multi circuit submission. This is a second attempt to implement this functionality because the first implementation involved a lot of code duplication with respect to the PennyLane baseline code.

Alex-Preciado commented:
Hi @radumarg , thank you for opening this new PR. @lillian542 will be reviewing it this week. In the meantime, could you please take a look at the formatting checks (black), CodeFactor, and codecov, as they’re currently not passing?

lillian542 commented on pennylane_ionq/device.py:
```diff
@@ -235,34 +380,164 @@ def _submit_job(self):
+            return result
+        except TypeError as e:
+            self._samples = None
+            e.args = ("When invoking QubitDevice estimate_probability() method after submitting circuits using any other method than QubitDevice.execute() method, current circuit index should be set via set_current_circuit_index() method pointing to index of the circuit which is targeted out of one or potentially several circuits submitted in a batch.", *e.args)
```
Is there something a user can do that makes this come up?

radumarg replied:
Removed, it was a bad idea. The feature I intended to capture here is now handled by CircuitIndexNotSetException which is now also properly tested.

lillian542 commented on pennylane_ionq/device.py:
```diff
@@ -235,34 +380,164 @@ def _submit_job(self):
+            dimension ``(shots,)`` or counts
+        """
+        try:
+            self._samples = self.generate_samples()
```
Why generate `self._samples` again, and then set them to `None` after calculating the result? Didn't both `execute` and `batch_execute` already do `self._samples = self.generate_samples()`? Of course if we clear them here, we can't use them for the next measurement, but why clear them here instead of clearing them in `reset`?

lillian542 reviewed (commented):
Thanks for the submission, and for making the changes requested on the last PR! 🚀 This looks like a very useful addition to the plugin. I’ve left some initial feedback and suggestions, have a look and see what you think!

Wrt the tests, the CodeCov bot has indicated all the uncovered lines, and ideally let’s go for full test coverage - but there are some places that I would say are more significant than others. I’ve commented on a few spots where I think it’s particularly important we add more comprehensive tests (i.e. more thorough than just what will satisfy CodeCov in terms of line coverage).

I’ll give it another pass once comments are addressed and the test coverage is more complete :)

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -151,6 +180,119 @@ def reset(self):
+            if self.shots is not None or any(
+                isinstance(m, sample_type) for m in circuit.measurements
+            ):
+                self._samples = self.generate_samples()
```
After inspecting the code I see now that on IonQDevice shots cannot be None, see the following error in IonQDevice init function: "The ionq device does not support analytic expectation values.". Consequently I removed completely the if clause.

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -234,34 +379,164 @@ def _submit_job(self):
+                provided, the entire shot range is treated as a single bin.
+            counts (bool): whether counts (``True``) or raw samples (``False``)
+                should be returned
+            circuit_index (int): index of circuit in case of batch circuit submission
```
fixed

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -234,34 +379,164 @@ def _submit_job(self):
+            result = super().sample(observable, shot_range, bin_size, counts)
+            self._samples = None
+            return result
+        except TypeError as e:
```
Removed, it was a bad ideea.

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -119,26 +136,38 @@ def __init__(
         self.error_mitigation = error_mitigation
         self.sharpen = sharpen
         self._operation_map = _GATESET_OPS[gateset]
+        self.histogram = None
+        self.histograms = None
```
removed self.histogram

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -119,26 +136,38 @@ def __init__(
+                "format": "ionq.circuit.v0",
+                "qubits": self.num_wires,
+                "circuits": [{"circuit": []} for _ in range(circuits_array_length)],
+                "gateset": self.gateset,
```
Implemented as suggested.

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -235,34 +380,164 @@ def _submit_job(self):
         """None or array[float]: Array of computational basis state probabilities. If
         no job has been submitted, returns ``None``.
         """
+        if self._current_circuit_index is not None:
+            self.histogram = self.histograms[self._current_circuit_index]
+
         if self.histogram is None:
             return None
 
```
I removed self.histogram. However, self._current_circuit_index remains defaulted to None instead 0. Here is the reason for that: when the number of circuits is larger than one and the user does not self._current_circuit_index an exception is raised notifying the user he must set a _current_circuit_index. If I would set _current_circuit_index 0  by default, a user which would use functions like dev.probability() or dev.sample() will always retrieve results for the first circuit which is incorrect and it would be very difficult for the user to notice he made a mistake. However 99% of the user will not use functions that are internal to device implementation like dev.probability() or dev.sample() because they would just submit a batch of circuits and then retrieve a batch of results and the whole _current_circuit_index issue will not be visible to them.

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -235,34 +380,164 @@ def _submit_job(self):
-        self.histogram = job.data.value
+        some_inner_value = next(iter(job.data.value.values()))
+        if isinstance(some_inner_value, dict):
+            self.histograms = []
+            for key in job.data.value.keys():
+                self.histograms.append(job.data.value[key])
```
This is extensively tested, see tests with "requires_api" in test_device.py These tests need an IonQ API key in the terminal where pytest is being run in order to run, but this is not my invention I just followed the example of some preexisting tests.

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -152,6 +181,119 @@ def reset(self):
+
+            # compute the required statistics
+            if self._shot_vector is not None:
+                result = self.shot_vec_statistics(circuit)
```
Added test_batch_execute_probabilities_with_shot_vector test

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -152,6 +181,119 @@ def reset(self):
+
+        # diagonalize observables
+        for operation in rotations:
+            self._apply_operation(operation, circuit_index)
```
Added test: test_batch_execute_expectation_value_with_diagonalization_rotations

radumarg commented:
@lillian542 The rework is complete. Review comments were implemented, tests were extended. Can you please make sure tests with requires_api flag are included in code coverage? Those tests need an IonQ API key in the test virtual environment in order to run. Note that using this is not my invention, I just took as an example some pre-existing testing code :)

lillian542 reviewed (commented):
Looking good 🎉  I left a couple more comments. At this point I'll find a second reviewer to have a look - we generally aim for 2 approvals for PRs adding new functionality.

radumarg commented on tests/test_device.py:
```diff
@@ -235,6 +250,319 @@ def test_backend_initialization(self, backend):
+        assert dev.tracker.history["resources"][0].shots.total_shots == 1024
+        assert len(dev.tracker.history["results"]) == 2
+
+    def test_not_recording_when_pennylane_tracker_not_active(self):
```
I do not understand why we need requires_api here, works just fine without it.

radumarg replied:
fixed as suggested

radumarg commented on tests/test_device.py:
```diff
@@ -235,6 +250,319 @@ def test_backend_initialization(self, backend):
+        dev.batch_execute([tape1])
+        assert dev.tracker.history == {}
+
+    def test_warning_on_empty_circuit(self):
```
I do not understand why we need requires_api here, works just fine without it.

radumarg replied:
fixed as suggested

radumarg commented on tests/test_device.py:
```diff
@@ -235,6 +250,319 @@ def test_backend_initialization(self, backend):
+    def test_batch_execute_logging_when_enabled(
+        self,
+        mock_logging_debug_method,
+        mock_logging_is_enabled_for_method,
```
I do not understand why we need requires_api here, test works just fine without it.

radumarg replied:
fixed as suggested

radumarg commented on tests/test_device.py:
```diff
@@ -235,6 +250,319 @@ def test_backend_initialization(self, backend):
         )
         assert dev.backend == backend
 
+    def test_recording_when_pennylane_tracker_active(self):
```
I do not understand why we need requires_api here, the test is working just fine without it.

lillian542 replied:
The CI for these 4 tests are failing with a "PermissionError: API key must be provided" error (see failures below), which I read as meaning the api key was needed, but maybe that's not the issue?

radumarg replied:
fixed as suggested

radumarg commented on pennylane_ionq/device.py:
```diff
@@ -235,42 +393,157 @@ def _submit_job(self):
+            Union[array[float], dict, list[dict]]: samples in an array of
+            dimension ``(shots,)`` or counts
+        """
+        self._samples = self.generate_samples()
+        result = super().sample(observable, shot_range, bin_size, counts)
+        self._samples = None
+        return result
+
+    def _measure(
+        self,
+        measurement: Union[SampleMeasurement, StateMeasurement],
+        shot_range=None,
+        bin_size=None,
+    ):
+        """Compute the corresponding measurement process depending on ``shots`` and the measurement
+        type. Overwrites the method in QubitDevice class to accomodate batch submission of circuits.
+        When invoking this method after submitting circuits using any other method than QubitDevice.execute()
+        method, current circuit index should be set via set_current_circuit_index() method pointing to index
+        of the circuit which is targeted out of one or potentially several circuits submitted in a batch.
+        This is not very good design but the alternative would be to create huge amounts of code duplication
+        with respect to the Pennylane base code.
+
+        Args:
+            measurement (Union[SampleMeasurement, StateMeasurement]): measurement process
+            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
+                to use. If not specified, all samples are used.
+            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
+                returns the measurement statistic separately over each bin. If not
+                provided, the entire shot range is treated as a single bin.
+
+        Raises:
+            ValueError: if the measurement cannot be computed
+
+        Returns:
+            Union[float, dict, list[float]]: result of the measurement
+        """
+        self._samples = self.generate_samples()
+        result = super()._measure(measurement, shot_range, bin_size)
+        self._samples = None
+        return result
```
These are needed, I'm afraid. The problem is that the user might decide to call these methods directly, outside of batch_execute for whatever reason.

lillian542 replied:
I understand your concern, and I’m not a big fan of storing `_samples` as a state on the device separately from using it 😅  In fact, the `_samples` attribute isn’t part of our new device API. But for plugins built on `QubitDevice` (the legacy API), it’s expected behaviour to need `_samples` to be set on the device before accessing `probability` or `sample`. The behaviour inherited from the parent class is consistent with our other PennyLane plugin devices.

radumarg replied:
I removed the two method override

radumarg commented:
@lillian542 Thanks for reviewing this change. I left a few questions regarding the latest set of comments. Please have a look.

lillian542 commented:
> Also I see that black formatter is failing but I am not sure why. When I run it locally it passes

@radumarg, when I run`black -l 100 pennylane_ionq`, it changes several split-line definitions into single lines. What version of `black` are you using? And is it possible that you are running it without the line length specification, and it's defaulting to something else?

mudit2812 commented on pennylane_ionq/device.py:
```diff
@@ -60,6 +73,18 @@
+    """
+
+    def __init__(self):
+        self.message = "Because multiple circuits have been submitted in this job, the index of the circuit \
+you want to acces must be first set via the set_current_circuit_index device method."
```
```suggestion
        self.message = (
            "Because multiple circuits have been submitted in this job, the index of the circuit "
            "you want to access must be first set via the set_current_circuit_index device method."
        )

```

radumarg replied:
fixed

mudit2812 commented on pennylane_ionq/device.py:
```diff
@@ -162,11 +308,16 @@ def operations(self):
         return set(self._operation_map.keys())
 
     def apply(self, operations, **kwargs):
+
+        "Implementation of QubitDevice abstract method apply."
```
```suggestion
        """Implementation of QubitDevice abstract method apply."""

```

radumarg replied:
fixed

mudit2812 commented on pennylane_ionq/device.py:
```diff
@@ -235,42 +393,124 @@ def _submit_job(self):
+        return prob_array
+
+    def estimate_probability(self, wires=None, shot_range=None, bin_size=None):
+        """Return the estimated probability of each computational basis state
+        using the generated samples. Overwrites the method in QubitDevice class
+        to accomodate batch submission of circuits. When invoking this method after
+        submitting circuits using any other method than QubitDevice.execute()
+        method, current circuit index should be set via set_current_circuit_index()
+        method to the circuit which is targeted out of one or potentially several
+        circuits submitted in a batch. This is not very good design but the
+        alternative would be to create huge amounts of code duplication with respect
+        to the Pennylane base code.
+
+        Args:
+            wires (Iterable[Number, str], Number, str, Wires): wires to calculate
+                marginal probabilities for. Wires not provided are traced out of one or the system.
+            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
+                to use. If not specified, all samples are used.
+            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
+                returns the measurement statistic separately over each bin. If not
+                provided, the entire shot range is treated as a single bin.
```
```suggestion
        """Return the estimated probability of each computational basis state
        using the generated samples. Overwrites the method in ``QubitDevice`` class
        to accommodate batch submission of circuits. When invoking this method after
        submitting circuits using any other method than ``QubitDevice.execute()``
        method, current circuit index should be set via ``set_current_circuit_index()``
        method to the circuit which is targeted out of one or potentially several
        circuits submitted in a batch. This is not very good design but the
        alternative would be to create huge amounts of code duplication with respect
        to the Pennylane base code.

        Args:
            wires (Iterable[Number, str], Number, str, Wires): wires to calculate
                marginal probabilities for. Wires not provided are traced out of one or the system.
            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
                to use. If not specified, all samples are used.
            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
                returns the measurement statistic separately over each bin. If not
                provided, the entire shot range is treated as a single bin.

```

mudit2812 commented on pennylane_ionq/device.py:
```diff
@@ -235,42 +393,124 @@ def _submit_job(self):
+        return prob_array
+
+    def estimate_probability(self, wires=None, shot_range=None, bin_size=None):
+        """Return the estimated probability of each computational basis state
+        using the generated samples. Overwrites the method in QubitDevice class
+        to accomodate batch submission of circuits. When invoking this method after
+        submitting circuits using any other method than QubitDevice.execute()
+        method, current circuit index should be set via set_current_circuit_index()
+        method to the circuit which is targeted out of one or potentially several
+        circuits submitted in a batch. This is not very good design but the
+        alternative would be to create huge amounts of code duplication with respect
+        to the Pennylane base code.
```
```suggestion
        """Return the estimated probability of each computational basis state
        using the generated samples. Overwrites the method in ``QubitDevice`` class
        to accommodate batch submission of circuits. When invoking this method after
        submitting circuits using any other method than ``QubitDevice.execute()``
        method, current circuit index should be set via ``set_current_circuit_index()``
        method to the circuit which is targeted out of one or potentially several
        circuits submitted in a batch. This is not very good design but the
        alternative would be to create huge amounts of code duplication with respect
        to the Pennylane base code.

```

radumarg replied:
fixed

mudit2812 commented on pennylane_ionq/device.py:
```diff
@@ -235,42 +393,124 @@ def _submit_job(self):
+        bin_size=None,
+        counts=False,
+    ):
+        """Return samples of an observable. Overwrites the method in QubitDevice class
+        to accomodate batch submission of circuits. When invoking this method after submitting
+        circuits using any other method than QubitDevice.execute() method, current circuit
+        index should be set via set_current_circuit_index() method pointing to index of the
+        circuit which is targeted out of one or potentially several circuits submitted in a batch.
+        This is not very good design but the alternative would be to create huge amounts of
+        code duplication with respect to the Pennylane base code.
```
```suggestion
        """Return samples of an observable. Overwrites the method in ``QubitDevice`` class
        to accommodate batch submission of circuits. When invoking this method after submitting
        circuits using any other method than ``QubitDevice.execute()`` method, current circuit
        index should be set via ``set_current_circuit_index()`` method pointing to index of the
        circuit which is targeted out of one or potentially several circuits submitted in a batch.
        This is not very good design but the alternative would be to create huge amounts of
        code duplication with respect to the Pennylane base code.

```

radumarg replied:
fixed

mudit2812 reviewed (changes requested):
Looks good. I mostly just suggested minor formatting/docs fixes. I agree with @lillian542 's comment about the generation of samples inside the `estimate_probability` and `sample` methods. We've resolved this concern for devices using the new API, but it would be a breaking change to use the current behaviour as per the old device API. For me, that is a blocker for an approval.

Other than that, the PR looks quite complete. The tests are very thorough. I'm happy to approve once the above comment and the test failures are resolved 🚀

radumarg commented:
@lillian542 and @mudit2812 Review comments were implemented as requested. I have added requires_api to failing tests, hopefully this time the test will pass.

mudit2812 reviewed (commented):
Looks like all the tests are passing now 😄 . I'm quite happy with the state of this PR, and am ready to approve once code coverage is fixed. We can ignore the coverage issue for the `logger.debug` line by adding a `# pragma: no cover` to that line.

radumarg commented:
@mudit2812 I have added the pragma as suggested. Also, I have removed checks for BasisState, QubitStateVector and StatePrep because these operations are not supported by IonQ devices so there is no way to test those lines. Regarding the code coverage: this now should be complete, however the code coverage tool does not take into account tests with requires_api which require an IonQ API key in order to run. For eample the code lines under "if self.tracker.active" are reported as uncovered but those lines are covered by this test: test_recording_when_pennylane_tracker_active The same applies for other code lines reported as uncovered.

mudit2812 reviewed (approved):
Thanks, I'm approving, and I've reached out to the rest of the team to figure out the best strategy to address the coverage issues.

lillian542 reviewed (approved):
Thanks for all your work here, this looks good! The tests aren’t running with the API token because this is on a fork of the repo - we’ll merge it to a new branch on the main repo to get those running.

I'm approving this before merging, it looks like everything is in order to me. If any issues come up once all the tests run, we’ll let you know; we can merge these changes to `master` as soon as the tests are passing with the API token there :)

radumarg commented:
Thanks, please let me know in case more tests will be necessary.
